### PR TITLE
IGNITE-21610 Upgrade tuples when inserting into indexes on a full state transfer

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-21610

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/index/IndexUpdateHandler.java [hunk 1]
```diff
@@ -65,10 +65,33 @@ public void addToIndexes(@Nullable BinaryRow binaryRow, RowId rowId, @Nullable L
         }
 
         for (TableSchemaAwareIndexStorage index : indexes(indexIds)) {
+            // TODO: IGNITE-21514 Handle index destruction
             index.put(binaryRow, rowId);
         }
     }
 
+    /**
+     * Adds a binary row to the index, if the tombstone then skips such operation.
```
Comment (rpuch) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/index/IndexUpdateHandler.java:74]:
```suggestion
     * Adds a binary row to the index, if it's a tombstone then skips such operation.
```

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/FullStateTransferIndexChooser.java [hunk 2]
```diff
@@ -317,4 +337,23 @@ private long findStoppingActivationTsBusy(int indexId, int toCatalogVersionInclu
...
+
+    private int tableVersionBusy(CatalogIndexDescriptor index, int catalogVersion) {
+        CatalogTableDescriptor table = catalogService.table(index.tableId(), catalogVersion);
+
+        assert table != null : "indexId=" + index.id() + ", tableId=" + index.tableId() + ", catalogVersion=" + catalogVersion;
+
+        return table.tableVersion();
+    }
+
+    private List<IndexIdAndTableVersion> toIndexIdAndTableVersionList(List<Integer> indexIds) {
```
Comment (rpuch) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/FullStateTransferIndexChooser.java:349]: How about calling this `enrichWithTableVersions`?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/IndexIdAndBinaryRow.java [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.table.distributed.raft.snapshot;
+
+import org.apache.ignite.internal.schema.BinaryRow;
+
+/** Container the index ID and {@link BinaryRow}. */
```
Comment (rpuch) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/IndexIdAndBinaryRow.java:22]:
```suggestion
/** Container for index ID and {@link BinaryRow}. */
```

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/IndexIdAndTableVersion.java [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.table.distributed.raft.snapshot;
+
+import org.apache.ignite.internal.tostring.S;
+
+/** Container the index ID and table schema version at index creation. */
```
Comment (rpuch) [inline modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/IndexIdAndTableVersion.java:22]:
```suggestion
/** Container for index ID and table schema version at index creation. */
```

**Review by rpuch — APPROVED** [review]

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/index/IndexUpdateHandler.java (hunk 1 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/index/IndexUpdateHandler.java:74]: apply it

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/FullStateTransferIndexChooser.java (hunk 2 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/FullStateTransferIndexChooser.java:349]: why not

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/IndexIdAndBinaryRow.java (hunk 3 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/IndexIdAndBinaryRow.java:22]: apply it

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/IndexIdAndTableVersion.java (hunk 4 above)
Comment (tkalkirill) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/raft/snapshot/IndexIdAndTableVersion.java:22]: apply it

**Pull request merged**